### PR TITLE
[BEAM-814] Dataflow: parallelize staging of files

### PR DESCRIPTION
**Pull request by dhalperi**
R: @lukecwik 

This is a WIP for the linked issue. Early feedback appreciated.

Decided to parallelize just hashing/compressing separately from uploading, because this should use a very tiny buffer per file and so can be parallelized nearly arbitrarily. In contrast, uploading uses a very large buffer and cannot be parallelized simply.

### CODE REVIEW

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 1]
```diff
@@ -144,48 +136,73 @@ static PackageAttributes createPackageAttributes(File classpathElement,
...
+      throw new RuntimeException("Error while staging packages", e.getCause());
+    } finally {
+      executorService.shutdown();
+    }
 
+    List<DataflowPackage> packages = new ArrayList<>();
+    int numUploaded = 0;
+    int numCached = 0;
+
+    for (PackageAttributes attributes : packageAttributes){
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: spacing between ){

Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: The real performance win can be found by parallelizing this loop.

We can get around the upload buffer size limit by doing an instanceof check and down casting to the GcsWriteChannel and setting the upload buffer size to be some sane limit like the smaller of 1mb or 1 byte greater than the file size.

Finally if we sort the file upload order to prioritize the largest files first, we can prevent the tail being dominated by a single large file that can't be split.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 2]
```diff
@@ -144,48 +136,73 @@ static PackageAttributes createPackageAttributes(File classpathElement,
...
-    ArrayList<DataflowPackage> packages = new ArrayList<>();
-
     if (stagingPath == null) {
       throw new IllegalArgumentException(
           "Can't stage classpath elements on because no staging location has been provided");
     }
 
-    int numUploaded = 0;
-    int numCached = 0;
+    ListeningExecutorService executorService = MoreExecutors.listeningDecorator(
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: Pass this in for testing or threadpool.

**Comment by dhalperi** [comment]
There are a lot of tests missing, but please take another look at the rewrite.

**Comment by dhalperi** [comment]
The memory suppression did not work. Investigation continues.

**Comment by dhalperi** [comment]
Because the channel is initialized deep inside GcsUtil before my code to set the option.

Pei's refactoring will expose this more easily, I think.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/GcsStager.java [hunk 3]
```diff
@@ -48,7 +52,14 @@ public static GcsStager fromOptions(PipelineOptions options) {
     if (windmillBinary != null) {
       filesToStage.add("windmill_main=" + windmillBinary);
     }
+    int uploadSizeBytes = options.getGcsUploadBufferSizeBytes() == null
+        ? 1024 * 1024
+        : Math.min(options.getGcsUploadBufferSizeBytes(), 1024 * 1024);
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/GcsStager.java]: Ensure the value is greater than 0

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java [hunk 4]
```diff
@@ -108,142 +100,232 @@ static PackageAttributes createPackageAttributes(File classpathElement,
...
+            + "copies to all workers. Having this many entries on your classpath may be indicative "
+            + "of an issue in your pipeline. You may want to consider trimming the classpath to "
+            + "necessary dependencies only, using --filesToStage pipeline option to override "
+            + "what files are being staged, or bundling several dependencies into one.",
+          classpathElements.size());
+    }
+
+    checkArgument(
+        stagingPath != null,
+        "Can't stage classpath elements on because no staging location has been provided");
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: Please update this awkward error message

Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java]: Done

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/GcsStager.java (hunk 3 above)
Comment (dhalperi) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/GcsStager.java]: Done

**Pull request closed without merging**